src/server.rs

```diff
@@ -665,9 +676,35 @@ pub fn generate_server_in_order_trait(
                 writeln!(out, "                    }}")?;
                 writeln!(out, "                    Err(val) => {{")?;
                 match err {
-                    // Note: for non-CLike errors we'd need to do an actual
-                    // reply here and then return Ok(()) to avoid invoking the
-                    // simple "return an integer" error path.
+                    syntax::Error::Complex(ty) => {
+                        // It might be surprising, but to return a complex error
+                        // we need to behave very much like the reply code
+                        // above: rather than returning `Err`, we need to
+                        // perform an actual `sys_reply` and then return `Ok` to
+                        // avoid triggering the reply handling code in the
+                        // generic dispatch loop.
+                        //
+                        // So, the fact that this error returns `Ok` is not a
+                        // bug.
+                        match op.encoding {
+                            syntax::Encoding::Hubpack => {
+                                writeln!(out, "                        match val {{")?;
+                                writeln!(out, "                            idol_runtime::RequestError::Fail(f) => {{")?;
+                                writeln!(out, "                                if let Some(fault) = f.into_fault() {{")?;
+                                writeln!(out, "                                    userlib::sys_reply_fault(rm.sender, fault);")?;
```

Comment:
If `f.into_fault()` returns `None`, does this fail to send a reply at all?

Reply:
It does, and that's in fact the intent and should maybe get a comment. Currently the only explanation is in `idol-runtime` [here](https://github.com/oxidecomputer/idolatry/blob/main/runtime/src/lib.rs#L65).